lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,271 @@
+  // Get loop information to eliminate loop exits from the dependencies
+  CFGLoopInfo li(domTree);
+
+  for (Block &block : funcReg.getBlocks()) {
+    // loop over all_control_deps_map to selectively pick the forward dependencies
+    for(size_t i = 0; i < all_control_deps_map[&block].size(); i++) {
+      Block* one_dep = all_control_deps_map[&block][i];
+      CFGLoop *loop = li.getLoopFor(one_dep);
+      if (loop == nullptr) {
+        // indicating that the one_dep is not inside any loop, so it must be a forward dependency
+        forward_control_deps_map[&block].push_back(one_dep);
+      } else {
+        // indicating that the one_dep is inside a loop, 
+        // to decide if it is a forward dep or not, compare it against all of the exits and latches of this loop
+        bool not_forward = false;
+        // check if one_dep is an exit of the loop
+        llvm::SmallVector<Block *> loop_exitBlocks;
+        loop->getExitingBlocks(loop_exitBlocks);
+        for(auto &loop_exit : loop_exitBlocks) {
+          if(loop_exit == one_dep) {
+            not_forward = true;
+            break;  // it is not a forward dependency so no need to contiue looping
+          }
+        }
+        // check if one_dep is a latch of the loop 
+        llvm::SmallVector<Block *> loop_latchBlocks;
+        loop->getLoopLatches(loop_latchBlocks);
+        for(auto &loop_latch : loop_latchBlocks) {
+          if(loop_latch == one_dep) {
+            not_forward = true;
+            break;  // it is not a forward dependency so no need to contiue looping
+          }
+        }
+
+        if (!not_forward) {
+          forward_control_deps_map[&block].push_back(one_dep);
+        }
+      }
+    }
+  }
+
+}
+
+void ControlDependenceAnalysis::returnAllControlDeps(mlir::Block* block, llvm::SmallVector<mlir::Block*, 4>& returned_all_control_deps) {
+  returned_all_control_deps = all_control_deps_map[block];
+}
+
+void ControlDependenceAnalysis::returnForwardControlDeps(mlir::Block* block, llvm::SmallVector<mlir::Block*, 4>& returned_forward_control_deps) {
+  returned_forward_control_deps = forward_control_deps_map[block];
+} 
+
+void ControlDependenceAnalysis::printAllBlocksDeps(mlir::func::FuncOp &funcOp) {
+  Region &funcReg = funcOp.getRegion();
+  llvm::errs() << "\n*********************************\n\n";
+  for (Block &block : funcReg.getBlocks()) {
+    block.printAsOperand(llvm::errs());
+    llvm::errs() << " is control dependent on: ";
+    for(size_t i = 0; i < all_control_deps_map[&block].size(); i++) {
+      all_control_deps_map[&block][i]->printAsOperand(llvm::errs());
+      llvm::errs() << ", ";
+    }
+    llvm::errs() << "\n";
+  }
+  llvm::errs() << "\n*********************************\n";
+}
+
+void ControlDependenceAnalysis::printForwardBlocksDeps(mlir::func::FuncOp &funcOp) {
+  Region &funcReg = funcOp.getRegion();
+  llvm::errs() << "\n*********************************\n\n";
```

Comment:
I think here we should print them as debug messages instead of error messages (you print an error message when there is an error).

See here:
https://github.com/EPFL-LAP/dynamatic/blob/6db10620892658395a2e88f199de26caf04f42ad/lib/Support/RTL.cpp#L281-L282